internal/cortex/querier/queryrange/query_range.go

```diff
@@ -217,9 +227,25 @@ func (prometheusCodec) MergeResponse(_ Request, responses ...Response) (Response
 	// we need to pass on all the headers for results cache gen numbers.
 	var resultsCacheGenNumberHeaderValues []string
 
+	var headers []*PrometheusResponseHeader
+	// merge headers
 	for _, res := range responses {
 		promResponses = append(promResponses, res.(*PrometheusResponse))
 		resultsCacheGenNumberHeaderValues = append(resultsCacheGenNumberHeaderValues, getHeaderValuesWithName(res, ResultsCacheGenNumberHeaderName)...)
+		for _, newHeader := range res.GetHeaders() {
+			found := false
+			for _, existing := range headers {
+				if existing.Name == newHeader.Name {
+					// if headers match, overwrite with the new header
+					existing = newHeader
```

Comment:
Should this be *existing = *newHeader ?

Also, below we sort the responses by response time and then the first non-null explanation is taken rather than the last one. Is consistency between these important?